modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -344,44 +409,69 @@
-          (when (a/<!! cancel-chan)
-            (log/debug "Received a message on the cancel channel; attempting to stop the BigQuery query execution")
-            (reset! cancel-requested? true) ; signal the page iteration fn to stop
-            (if-not (or (future-cancelled? res-fut) (future-done? res-fut))
-              ;; somehow, even the FIRST page hasn't come back yet (i.e. the .query call above), so cancel the future to
-              ;; interrupt the thread waiting on that response to come back
-              ;; unfortunately, with this particular overload of .query, we have no access to (nor the ability to control)
-              ;; the jobId, so we have no way to use the BigQuery client to cancel any job that might be running
-              (future-cancel res-fut)
-              (when (future-done? res-fut) ; canceled received after it was finished; may as well return it
-                @res-fut)))))
-      @res-fut)
-    (catch java.util.concurrent.CancellationException _e
-      (throw (ex-info (tru "Query cancelled") {:sql sql :parameters parameters ::cancelled? true})))
-    (catch BigQueryException e
-      (if (.isRetryable e)
-        (throw (ex-info (tru "BigQueryException executing query")
-                        {:retryable? (.isRetryable e), :sql sql, :parameters parameters}
-                        e))
-        (throw-invalid-query e sql parameters)))
-    (catch Throwable e
-      (throw-invalid-query e sql parameters))))
+  (let [request (doto (QueryJobConfiguration/newBuilder sql)
+                  ;; if the query contains a `#legacySQL` directive then use legacy SQL instead of standard SQL
+                  (.setUseLegacySql (str/includes? (u/lower-case-en sql) "#legacysql"))
+                  (bigquery.params/set-parameters! parameters)
+                  ;; .setMaxResults is very misleading; it's actually the page size, and it only takes
+                  ;; effect for RPC (a.k.a. "fast") calls
+                  ;; there is no equivalent of .setMaxRows on a JDBC Statement; we rely on our middleware to stop
+                  ;; realizing more rows as per the maximum result size
+                  (.setMaxResults *page-size*))
+        ;; as long as we don't set certain additional QueryJobConfiguration options, our queries *should* always be
+        ;; following the fast query path (i.e. RPC)
+        ;; check out com.google.cloud.bigquery.QueryRequestInfo.isFastQuerySupported for full details
+        done-chan  (a/promise-chan)
+        res-fut    (future
+                     (log/trace "BigQuery exec thread sending query job")
+                     (try
+                       (let [result (.query client (.build request) (u/varargs BigQuery$JobOption))]
+                         (log/trace "BigQuery request finished successfully; sending the query to the channel")
+                         (let [pushed? (a/>!! done-chan [:done result])]
```

Comment:
I think you still should be calling `a/close!` on this even if it's a promise channel... that's another reason to prefer a plain `promise` IMO, you don't have to worry about remembering to close it